### PR TITLE
Escape strings in HTML output (XSS fix)

### PR DESCRIPTION
**Pull request by zmetcalf**
### Description

- Added XSS to HTML and JSON output
- Bumped Version of EJS for recent security vulnerability

The original issue was around output of the HTML, but the logger still produced XSS issues in JSON. I included a fix for that too. I did not do a sanitize function in `lib/data-builder.js` because it threw off the sanitation in the XML portion. I included tests for HTML, XML and JSON. 

#### Related issues

- connect to #66 

### Checklist

- [X] New tests added or existing tests modified to cover all changes
- [X] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

File: .gitignore [hunk 1]
```diff
@@ -31,3 +31,6 @@ node_modules
 
 # Optional REPL history
 .node_repl_history
+
+# Npm Lock File
+package-lock.json
```
Comment (bajtos) [inline .gitignore]: We are disabling package-lock completely via `.npmrc`, see e.g. https://github.com/strongloop/loopback-next/blob/master/.npmrc

Please revert this change and add `.npmrc` file instead.

File: lib/send-json.js [hunk 2]
```diff
@@ -5,8 +5,14 @@
 
 'use strict';
 
+var escapeHtml = require('escape-html');
+
 module.exports = function sendJson(res, data) {
-  var content = JSON.stringify({error: data});
+  var content = JSON.stringify({error: data}, function(key, value) {
+    return typeof value  === 'string' && key !== 'stack' ?
+      escapeHtml(value) :
```
Comment (bajtos) [inline lib/send-json.js]: I am afraid this does not make any sense to me. The JSON is a transport format that can be used in different ways - store it in SQL (think of a table storing all failed responses for future inspection), render it as HTML from a single-page app, etc. It's the responsibility of the code consuming this data to apply any escaping needed, depending on how it works with that data.

Please revert this change.

File: package.json [hunk 3]
```diff
@@ -19,7 +19,8 @@
   "dependencies": {
     "accepts": "^1.3.3",
     "debug": "^2.2.0",
-    "ejs": "^2.4.2",
+    "ejs": "^2.5.5",
```
Comment (bajtos) [inline package.json]: AFAICT, the latest `ejs` version is `2.5.7`. Please use that version instead of `2.5.5`.

File: test/handler.test.js [hunk 4]
```diff
@@ -472,6 +498,36 @@ describe('strong-error-handler', function() {
           done);
     });
 
+    it('sanitizes response to prevent XSS', function(done) {
+      var error = new ErrorWithProps({
```
Comment (bajtos) [inline test/handler.test.js]: Use `const`, see http://loopback.io/doc/en/contrib/style-guide.html#variable-declarations

File: test/handler.test.js [hunk 5]
```diff
@@ -472,6 +498,36 @@ describe('strong-error-handler', function() {
...
+      });
+      givenErrorHandlerForError(error, {debug: false});
+      requestHTML()
+        .end(function(err, res) {
+          expect(res.statusCode).to.eql(404);
+          var body = res.error.text;
+          expect(body).to.match(
+            /some details&lt;img onerror=alert\(1\) src=a&gt;/
+          );
+          expect(body).to.not.match(
```
Comment (bajtos) [inline test/handler.test.js]: Please remove these `not.match` assertions. There are other tests that have already verified that sensitive information is excluded. Let's keep this new test focused on HTML escaping only. Please remove `details` and `extra` from the error object created at the beginning of this test too, since they won't be used anymore.

File: test/handler.test.js [hunk 6]
```diff
@@ -472,6 +498,36 @@ describe('strong-error-handler', function() {
           done);
     });
 
+    it('sanitizes response to prevent XSS', function(done) {
```
Comment (bajtos) [inline test/handler.test.js]: Please make this test name more descriptive - e.g. `HTML-escapes response fields to prevent XSS in production mode`.

File: test/handler.test.js [hunk 7]
```diff
@@ -594,6 +650,35 @@ describe('strong-error-handler', function() {
         });
     });
 
+    it('sanitizes response to prevent XSS', function(done) {
```
Comment (bajtos) [inline test/handler.test.js]: Please make this test name more descriptive - e.g. `HTML-escapes all 4xx response properties in production mode`.

File: package.json [hunk 8]
```diff
@@ -19,7 +19,8 @@
...
+    "escape-html": "^1.0.3",
```
Comment (bajtos) [inline package.json]: This won't be needed after the changes in JSON output are reverted, please remove.

**Comment by bajtos** [comment]
@slnode ok to test

**Comment by bajtos** [comment]
> Added XSS to HTML and JSON output 73299cb

We will need to fix the commit message to follow our style (see [here](http://loopback.io/doc/en/contrib/git-commit-messages.html)), but that's probably best left for the end, once the code changes are good.

File: lib/send-json.js (hunk 2 above)
Comment (zmetcalf) [reply lib/send-json.js]: @bajtos This may not be the right fix, but it is the type of fix that we actually wanted make because it is a low level security issue by pushing out potentially hazardous text to the frontend (I just did the HTML to be nice). I agree it is the developer's responsibility to ensure proper HTML escaping, but shouldn't this library (or loopback.io) be cautious of what it returns?

Comment (bajtos) [reply lib/send-json.js]: In the world of API Economy, there are more and more API consumers that are not HTML websites. Consider for example a conversational bot invoking LoopBack APIs in background. Would you like to see HTML-escaped error messages in your Skype chat session because LoopBack is assuming all JSON consumers are web apps? If this example looks too silly, then consider a microservice invoking a LoopBack backend and reporting errors returned by LoopBack to a text log file. I would not want to see HTML-escaped text there, because it's unnecessarily difficult to read.

Don't get me wrong, I understand that in your particular case it may be helpful (or at least tempting) to escape string values as early as possible. But I still think escaping JSON values is not really solving the security problem. If you don't trust your HTML front-end is correctly escaping string values, then you should be HTML-escaping strings in all JSON responses, including those fetching data from your database. And if your database is SQL-based and we apply the same logic, then you need to SQL-encode all input values coming from your HTML front-ends too, to make sure they don't accidentally trigger SQL-injection attack in your backend.

Anyways, I don't mind supporting such setup by allowing strong-error-handler users to provide a custom function for producing JSON response bode, there is already a work in progress in #67.

However, the default behavior of strong-error-handler should stay as it is now, i.e. no HTML-escaping in JSON responses.

**Comment by zmetcalf** [comment]
@bajtos I have updated the commit with your initial round of updates. Please advise any additional changes.

**Comment by bajtos** [comment]
Landed 🎉  Thank you for the contribution! ❤️
